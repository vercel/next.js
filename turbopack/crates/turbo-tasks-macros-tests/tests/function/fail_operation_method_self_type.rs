#![allow(dead_code)]
#![feature(arbitrary_self_types)]
#![feature(arbitrary_self_types_pointers)]

use turbo_tasks::{OperationVc, Vc};

#[turbo_tasks::value]
struct Foobar;

#[turbo_tasks::value_impl]
impl Foobar {
    #[turbo_tasks::function(operation)]
    fn arbitrary_self_type(self: OperationVc<Self>) -> Vc<()> {
        Vc::cell(())
    }
}

fn main() {}
