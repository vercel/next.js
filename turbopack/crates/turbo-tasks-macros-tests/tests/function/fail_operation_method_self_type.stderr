error: methods taking `self` are not supported with `operation`
  --> tests/function/fail_operation_method_self_type.rs:13:28
   |
13 |     fn arbitrary_self_type(self: OperationVc<Self>) -> Vc<()> {
   |                            ^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `fn(OperationVc<Foobar>) -> Vc<()> {Foobar::arbitrary_self_type_turbo_tasks_function_inline}: turbo_tasks::task::function::IntoTaskFnWithThis<_, _, _>` is not satisfied
  --> tests/function/fail_operation_method_self_type.rs:10:1
   |
10 | #[turbo_tasks::value_impl]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ unsatisfied trait bound
   |
   = help: the trait `turbo_tasks::task::function::TaskFnInputFunctionWithThis<_, _, _>` is not implemented for fn item `fn(OperationVc<Foobar>) -> Vc<()> {Foobar::arbitrary_self_type_turbo_tasks_function_inline}`
   = note: required for `fn(OperationVc<Foobar>) -> Vc<()> {Foobar::arbitrary_self_type_turbo_tasks_function_inline}` to implement `turbo_tasks::task::function::IntoTaskFnWithThis<_, _, _>`
note: required by a bound in `NativeFunction::new_method`
  --> $WORKSPACE/turbopack/crates/turbo-tasks/src/native_function.rs
   |
   |     pub fn new_method<Mode, This, Inputs, I>(
   |            ---------- required by a bound in this associated function
...
   |         I: IntoTaskFnWithThis<Mode, This, Inputs>,
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `NativeFunction::new_method`
   = note: this error originates in the attribute macro `turbo_tasks::value_impl` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0307]: invalid `self` parameter type: `OperationVc<Foobar>`
  --> tests/function/fail_operation_method_self_type.rs:13:34
   |
13 |     fn arbitrary_self_type(self: OperationVc<Self>) -> Vc<()> {
   |                                  ^^^^^^^^^^^^^^^^^
   |
   = note: type of `self` must be `Self` or a type that dereferences to it
   = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)
